dds/idl/marshal_generator.cpp

```diff
@@ -2378,29 +2378,25 @@ bool marshal_generator::gen_struct(AST_Structure* node,
     code.push_back("if (!strm.read_delimiter(total_size)) {");
     code.push_back("  return false;");
     code.push_back("}");
+    if (not_final && repr.xcdr2) {
+      be_global->impl_ <<
+        "  size_t start_pos = strm.pos();\n";
+    }
     generate_dheader_code(code, not_final);
 
     if (may_be_parameter_list) {
-      if (repr.xcdr2) {
-        /**
-         * We don't have a PID marking the end in XCDR2 parameter lists, but we
-         * have the size, so we need to stop after we hit this offset.
-         *
-         * TODO(iguessthislldo): Replace this with a cleaner mechanism where
-         * the Serializer keeps track of the stream offset.
-         * https://github.com/objectcomputing/OpenDDS/pull/1722#discussion_r447056830
-         */
-        be_global->impl_ <<
-          "  const char* end_of_fields = strm.pos_rd() + total_size;\n";
-      }
       be_global->impl_ <<
         "  unsigned member_id;\n"
         "  size_t field_size;\n"
         "  while (true) {\n";
 
       if (repr.xcdr2) {
+        /**
+         * We don't have a PID marking the end in XCDR2 parameter lists, but we
+         * have the size, so we need to stop after we hit this offset.
+         */
         be_global->impl_ << "\n"
-          "    if ((!strm.pos_rd() || strm.pos_rd() >= end_of_fields)";
+          "    if (strm.pos() - start_pos >= total_size + uint32_cdr_size";
```

Comment:
```suggestion
          "    if (strm.pos() - start_pos >= total_size";
```

If the `start_pos` is set after dheader, then you don't have to account for it here.